contracts/mainnet/connectors/fluid-vaults/main.sol

```diff
@@ -0,0 +1,299 @@
+     *              7 - token1 borrow amount
+     *              8 - token1 payback amount
+     */
+    function operatePerfectWithIds(
+        address vaultAddress_,
+        uint256 nftId_,
+        int256 perfectColShares_,
+        int256 colToken0MinMax_, // if +, max to deposit, if -, min to withdraw
+        int256 colToken1MinMax_, // if +, max to deposit, if -, min to withdraw
+        int256 perfectDebtShares_,
+        int256 debtToken0MinMax_, // if +, min to borrow, if -, max to payback
+        int256 debtToken1MinMax_, // if +, min to borrow, if -, max to payback
+        uint256 getNftId_,
+        uint256[] memory setIds_
+    )
+        external
+        payable
+        returns (string memory _eventName, bytes memory _eventParam)
+    {
+        nftId_ = getUint(getNftId_, nftId_);
+        IVaultT4 vaultT4_ = IVaultT4(vaultAddress_);
+        IVaultT4.ConstantViews memory vaultT4Details_ = vaultT4_
+            .constantsView();
+        uint256 ethAmount_;
+
+        if (perfectColShares_ > 0) {
+            (ethAmount_, ) = _handleDeposit(
+                vaultT4Details_.supplyToken.token0 == getEthAddr(),
+                false,
+                vaultAddress_,
+                vaultT4Details_.supplyToken.token0,
+                colToken0MinMax_
+            );
+
+            (ethAmount_, ) = _handleDeposit(
+                vaultT4Details_.supplyToken.token1 == getEthAddr(),
+                false,
+                vaultAddress_,
+                vaultT4Details_.supplyToken.token1,
+                colToken1MinMax_
+            );
+        }
+
+        // Payback
+        if (perfectDebtShares_ < 0) {
+            ethAmount_ = _handlePayback(
+                vaultT4Details_.borrowToken.token0 == getEthAddr(),
+                false,
+                vaultT4Details_.borrowToken.token0,
+                0, // Not needed since isMin is always false
+                debtToken0MinMax_,
+                vaultAddress_
+            );
+
+            ethAmount_ = _handlePayback(
+                vaultT4Details_.borrowToken.token1 == getEthAddr(),
+                false,
+                vaultT4Details_.borrowToken.token1,
+                0, // Not needed since isMin is always false
+                debtToken1MinMax_,
+                vaultAddress_
+            );
+        }
+
+        int256[] memory r_;
+
+        (nftId_, r_) = vaultT4_.operatePerfect(
+            nftId_,
+            perfectColShares_,
+            colToken0MinMax_,
+            colToken1MinMax_,
+            perfectDebtShares_,
+            debtToken0MinMax_,
+            debtToken1MinMax_,
+            address(this)
+        );
+
+        setUint(setIds_[0], nftId_);
+
+        _handleOperatePerfectSetIds(r_[2], setIds_[1], setIds_[2]);
+        _handleOperatePerfectSetIds(r_[3], setIds_[3], setIds_[4]);
+        _handleOperatePerfectSetIds(r_[5], setIds_[5], setIds_[6]);
+        _handleOperatePerfectSetIds(r_[6], setIds_[7], setIds_[8]);
+
```

Comment:
Would also be nice to make the approval 0 as tokens like USDT starts to fail if approval is not 0.

Reply:
This applies on both the functions

Reply:
Made approval 0 in operate perfect.